README.md
# mb.flipText 1.1

__An open source jQuery component to flip your text vertically.__

![mb.flipText](http://pupunzi.com/gitHub/mb.flipText.jpg)

## [go to the demo](http://pupunzi.com/#mb.components/mb.flipText/flipText.html)
## [go to the doc](http://wiki.github.com/pupunzi/jquery.mb.flipText/)
## [go to the blog](http://pupunzi.open-lab.com/mb-jquery-components/jquery-mb-flipText/)


[jquery.mb.components](http://pupunzi.com/), another way of thinking the web


inc/changeLog.txt
jquery.mb.flipText

1.1
fixed some bug and improved performance with several istances.

1.0 first release.

inc/jquery.mb.flipText.pack.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: info@pupunzi.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * jQuery.mb.components: jquery.mb.flipText
 * version: 1.0- 4-dic-2009 - 55
 * © 2001 - 2009 Matteo Bicocchi (pupunzi), Open Lab
 *
 */
eval(function(p,a,c,k,e,r){e=function(c){return(c<a?'':e(parseInt(c/a)))+((c=c%a)>35?String.fromCharCode(c+29):c.toString(36))};if(!''.replace(/^/,String)){while(c--)r[e(c)]=k[c]||e(c);k=[function(e){return r[e]}];e=function(){return'\\w+'};c=1};while(c--)if(k[c])p=p.replace(new RegExp('\\b'+e(c)+'\\b','g'),k[c]);return p}('(p($){6 n=$.t.P;Q.G.H=p(){I 3.R(p(){6 a=$(3);6 b=a.q();a.q(b.8(/&/g,\'&1f;\').8(/</g,\'&1g;\').8(/>/g,\'&1h;\').8(/\'/g,S("\'")).8(/"/g,S(\'"\')))})};$.J={1i:"1j 1k",1l:"1.0",T:p(m){I 3.R(p(){6 h=$(3).u(U)[1]+"V";6 w=$(3).u(U)[0]+"V";r(!n)$(3).H();6 a=$(3).B();6 b="";6 c=$(3).7("K-C")?$(3).7("K-C"):"#1m";6 d=o($(3).7(\'9-D\'))>0?o($(3).7(\'9-D\')):14;6 e=$(3).7(\'9-E\')?$(3).7(\'9-E\').8(/\\\'/g,\'\').8(/"/g,\'\'):"W";6 f=$(3).7(\'C\')?$(3).7(\'C\'):"#1n";r($.t.P){r(!m)$(3).7({\'X-Y\':\'Z-10\',v:h,1o:\'1p() 1q("") \',L:"11"}).7(\'9-1r\',\'1s\');b=$("<M F=\'X-Y: Z-10; L:11; v:"+h+"; z:"+w+"\'>"+a+"</M>")}1t{6 g="-12";6 i="1u";6 j=0;6 k=$.t.1v?o(w)-(o(w)/4):$.t.13?5:"0";r(m){j=$.t.13?5:0;g="12, "+((o(w)/2)-j)+", "+o(w)/2;i="1w"}b=$("<15 17=\'1x\' F=\'v:"+h+"; z:"+w+";\' 1y=\'18/A+19\' 1a=\'1a:18/A+19,"+"<A 1z=\\"1A://1B.1C.1D/1E/A\\">"+"<1F x=\\"0\\" y=\\"0\\" z=\\""+w+"\\" v=\\""+h+"\\" 1b=\\""+c+"\\" 1G=\\"1H\\"/>"+"<q  x=\\"-"+j+"\\" y=\\""+(k)+"\\" 9-E=\\""+e+"\\"  1b=\\""+f+"\\" 9-D=\\""+d+"\\"  F=\\"q-1I: "+i+"; "+"1J-1K: 1L\\" 1M=\\"1N("+g+")\\" q-1O=\\"1P\\">"+a+"</q></A>\'></15>"+"<1Q 17=\'1R\' F=\'s:N;1c:0;1S:0;z:O%;v:O%;K:1T\'/>"+"")}6 l=$(3).7("s")!="N"||$(3).7("s")!="1U"?"1V":$(3).7("s");$(3).B(b).7({s:l,z:w})})},u:p(a){r(!a&&!n)$(3).H();6 b=$(3).B();6 c=o($(3).7(\'9-D\'));6 d=$(3).7(\'9-E\').8(/\'/g,\'\').8(/"/g,\'\');r(d==1W)d="W";6 e=$("<M/>").7({s:"N",1c:-O,L:"1X",1Y:c,1Z:d});e.B(b);$("20").21(e);6 h=(e.1d()!=0?e.1d():(16+b.22*c*.23));6 w=(e.1e()!=0?e.1e()+5:24);e.25();I[w,h]}};$.G.26=$.J.T;$.G.u=$.J.u})(Q);',62,131,'|||this|||var|css|replace|font|||||||||||||||parseInt|function|text|if|position|browser|getFlipTextDim|height||||width|svg|html|color|size|family|style|fn|encHTML|return|mbflipText|background|whiteSpace|span|absolute|100|msie|jQuery|each|escape|flipText|true|px|Arial|writing|mode|tb|rl|nowrap|90|safari||object||class|image|xml|data|fill|top|outerWidth|outerHeight|amp|lt|gt|author|Matteo|Bicocchi|version|fff|000|filter|fliph|flipv|weight|normal|else|end|opera|start|flip_label|type|xmlns|http|www|w3|org|2000|rect|stroke|none|anchor|dominant|baseline|hanging|transform|rotate|rendering|optimizeSpeed|div|pointer|left|transparent|fixed|relative|undefined|noWrap|fontSize|fontFamily|body|append|length|60|50|remove|mbFlipText'.split('|'),0,{}))

inc/jquery.mb.flipText.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: info@pupunzi.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * jQuery.mb.components: jquery.mb.flipV
 * version: 1.1
 * © 2001 - 2009 Matteo Bicocchi (pupunzi), Open Lab
 *
 */

(function($) {
  var isIE=$.browser.msie;
  jQuery.fn.encHTML = function() {
    return this.each(function(){
      var me   = $(this);
      var html = me.text();
      me.text(html.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/'/g, escape("'")).replace(/"/g,escape('"')));
    });
  };
  $.mbflipText= {
    author:"Matteo Bicocchi",
    version:"1.1",
    flipText:function(tb){
      var UTF8encoded=$("meta[http-equiv=Content-Type]").attr("content").indexOf("utf-8")>-1;
      return this.each(function(){
        var el= $(this);
        var h="";
        var w="";

        if(!isIE && !UTF8encoded) el.encHTML();
        var txt= el.html();
        var label="";
        var bgcol=(el.css("background-color") && el.css("background-color") != "rgba(0, 0, 0, 0)") ? el.css("background-color"):"#fff";
        var fontsize= parseInt(el.css('font-size'))>0?parseInt(el.css('font-size')):14;
        var fontfamily=el.css('font-family')?el.css('font-family').replace(/\'/g, '').replace(/"/g,''):"Arial";
        var fontcolor=el.css('color')? el.css('color'):"#000";

        if ($.browser.msie){
          if(!tb) el.css({'writing-mode': 'tb-rl', height:h, filter: 'fliph() flipv("") ', whiteSpace:"nowrap"}).css('font-weight', 'normal');
          label=$("<span style='writing-mode: tb-rl; whiteSpace:nowrap; height:"+h+"; width:"+w+"'>"+txt+"</span>");
        }else{

          h=el.getFlipTextDim(true)[1]+"px";
          w=el.getFlipTextDim(true)[0]+"px";

          var rot="-90";
          var ta="end";
          var xFix=0;
          var yFix=$.browser.opera ? parseInt(w)-(parseInt(w)/4): $.browser.safari?0:0;
          if (tb){
            xFix= $.browser.safari?(fontsize/4):0;
            rot="90, "+((parseInt(w)/2)-xFix)+", "+parseInt(w)/2;
            ta="start";
          }
          var onClick= el.attr("onclick") || el.attr("href");
          var clickScript= onClick?"<div class='pointer' style='position:absolute;top:0;left:0;width:100%;height:100%;background:transparent'/>":"";

          label=$("<object class='flip_label' style='height:"+h+"; width:"+w+";' type='image/svg+xml' data='data:image/svg+xml; charset=utf-8 ," +
                  "<svg xmlns=\"http://www.w3.org/2000/svg\">" +
                  "<rect x=\"0\" y=\"0\" width=\""+w+"\" height=\""+h+"\" fill=\""+bgcol+"\" stroke=\"none\"/>"+
                  "<text  x=\"-"+xFix+"\" y=\""+yFix+"\" font-family=\""+fontfamily+"\"  fill=\""+fontcolor+"\" font-size=\""+fontsize+"\"  style=\"text-anchor: "+ta+"; " +
                  "dominant-baseline: hanging\" transform=\"rotate("+rot+")\" text-rendering=\"optimizeSpeed\">"+txt+"</text></svg>'></object>" +
                  clickScript +
                  "");
        }
        var wrapper= onClick ? $("<div/>").css("position","relative"): $("");
        var cssPos= el.wrap(wrapper).css("position")!="absolute" || el.css("position")!="fixed"  ?"relative" : el.css("position");
//        var cssPos= el.css("position")!="absolute" || el.css("position")!="fixed"  ?"relative" : el.css("position");

        el.html(label).css({position:cssPos, width:w});
      });
    },
    getFlipTextDim:function(enc){
      var el= $(this);
      if(!enc && !isIE) el.encHTML();
      var txt= el.html();
      var fontsize= parseInt(el.css('font-size'));
      var fontfamily=el.css('font-family').replace(/'/g, '').replace(/"/g,'');
      if (fontfamily==undefined) fontfamily="Arial";
      var placeHolder=$("<span/>").css({position:"absolute",top:-100, whiteSpace:"noWrap", fontSize:fontsize, fontFamily: fontfamily});
      placeHolder.html(txt);
      $("body").append(placeHolder);
      var h = (placeHolder.outerWidth()!=0?placeHolder.outerWidth():(16+txt.length*fontsize*.60));
      var w = (placeHolder.outerHeight()!=0?placeHolder.outerHeight()+5:50);
      placeHolder.remove();
      return [w,h];
    }
  };
  $.fn.mbFlipText=$.mbflipText.flipText;
  $.fn.getFlipTextDim=$.mbflipText.getFlipTextDim;

})(jQuery);